src/idpi/operators/stencils.py

```diff
@@ -0,0 +1,115 @@
+        -1: slice(-2),
+        0: slice(1, -1),
+        1: slice(2, None),
+    },
+    "half": {
+        -1: slice(-1),
+        0: slice(None),
+        1: slice(1, None),
+    },
+}
+
+
+class Field(Protocol):
+    def dx(self) -> xr.DataArray:
+        ...
+
+    def dy(self) -> xr.DataArray:
+        ...
+
+    def dz(self) -> xr.DataArray:
+        ...
+
+
+class PaddedField:
+    def __init__(self, field: xr.DataArray, vert: Vert = "generalVerticalLayer"):
+        self.indexes = field.indexes
+        self.vert = vert
+        tmp = field.pad({vert: 1}, mode="edge")
+        self.field = tmp.pad({dim: 1 for dim in ("x", "y")}, constant_values=np.nan)
+
+    def __getitem__(self, indices: tuple[int, int, int]) -> xr.DataArray:
+        i, j, k = indices
+        s = STENCILS["full"]
+        sel = {"x": s[i], "y": s[j], self.vert: s[k]}
+        return self.field[sel].assign_coords(self.indexes)
+
+    def dx(self) -> xr.DataArray:
+        return 0.5 * (self[1, 0, 0] - self[-1, 0, 0])
+
+    def dy(self) -> xr.DataArray:
+        return 0.5 * (self[0, 1, 0] - self[0, -1, 0])
+
+    def dz(self) -> xr.DataArray:
+        result = self[0, 0, 1] - self[0, 0, -1]
+        result[{self.vert: slice(1, -1)}] *= 0.5
+        return result
+
+
+class StaggeredField:
+    """Should have one more element along the staggered dimension."""
+
+    def __init__(self, field: xr.DataArray):
+        if field.indexes:
+            field = field.drop_indexes(field.indexes.keys())
+        self.field = field
+        self.padded = PaddedField(destagger_z(field))
+
+    def __getitem__(self, indices: tuple[int, int, int]) -> xr.DataArray:
+        i, j, k = indices
+        s = STENCILS["half"]
+        return self.field[{"x": s[i], "y": s[j], "generalVertical": s[k]}]
+
+    def dx(self):
+        return self.padded.dx()
+
+    def dy(self):
+        return self.padded.dy()
+
+    def dz(self):
+        return (self[0, 0, 1] - self[0, 0, -1]).rename(
+            generalVertical="generalVerticalLayer"
+        )
+
+
+def destagger_z(field: xr.DataArray) -> xr.DataArray:
+    z = "generalVertical"
+    result = 0.5 * (field + field.shift({z: -1})).dropna(z, how="all")
+    return result.rename({z: "generalVerticalLayer"})
+
+
+class TotalDiff:
+    def __init__(self, dlon: float, dlat: float, hhl: xr.DataArray):
+        self.dlon = dlon
+        self.dlat = dlat
+        hhl_pad = PaddedField(hhl, "generalVertical")
+        dh_dx = destagger_z(hhl_pad.dx())  # order is important
+        dh_dy = destagger_z(hhl_pad.dy())  # diff then destagger
+
+        dim = "generalVertical"
+        rename_kw = {dim: "generalVerticalLayer"}
+        self.sqrtg_r_s = -1 / hhl.diff(dim=dim, label="lower").rename(rename_kw)
+        self.dzeta_dlam = self.sqrtg_r_s / dlon * dh_dx
+        self.dzeta_dphi = self.sqrtg_r_s / dlat * dh_dy
+
+    def d_dlam(self, field: Field) -> xr.DataArray:
+        return field.dx() / self.dlon + field.dz() * self.dzeta_dlam
```

Comment:
I am not sure I understand why dx, dy were coded different than dz. 
I think the delta function is `(self[1, 0, 0] - self[-1, 0, 0])` irrespective of the axis, instead of `0.5 * (self[1, 0, 0] - self[-1, 0, 0])`. 
In the the x dimension here we would have: 
`delta(field, 'x') / (2*self.dlon)`
very similar for the z dimension but instead of 2*(hhl[0,0,1]-hhl[0,0,0], we need to use the geometrical coefficient dzeta_dlam. So the partial derivatives are different, but this is due to the differences between z and horizontal dimension, and not related to staggering, isnt?
So like that the delta function can be coded invariant wrt the axis.

Reply:
The idea was to make `dx, dy, dz` return the derivatives of the field with respect to unit length coordinates in the 3 directions respectively. Then, multiply those by the grid deformation factors. I am aware that there is some inconsistency on the naming but I tried to stick to the naming in fieldextra. I'm a bit unclear on the naming conventions so perhaps I have introduced some further inconsistencies. Also, note that the `self[1, 0, 0]` returns different stencils whether you're looking at `PaddedField` or `StaggeredField` and the first has a different treatment of the vertical as well. This is done to maintain consistency with the vertical boundary management in fieldextra. I agree that these are all potential sources of confusion but I fail to see how changing the unit length to an arbitrary length will help reduce the confusion.